src/Symfony/Bridge/Doctrine/DependencyInjection/CompilerPass/RegisterEventListenersAndSubscribersPass.php

```diff
@@ -93,15 +95,22 @@ private function addTaggedListeners(ContainerBuilder $container)
                 if (!isset($this->connections[$con])) {
                     throw new RuntimeException(sprintf('The Doctrine connection "%s" referenced in service "%s" does not exist. Available connections names: %s', $con, $id, implode(', ', array_keys($this->connections))));
                 }
-
-                if ($lazy = !empty($tag['lazy'])) {
-                    $taggedListenerDef->setPublic(true);
-                }
+                $listenerClosures[$con][$id] = new ServiceClosureArgument(new Reference($id));
 
                 // we add one call per event per service so we have the correct order
-                $this->getEventManagerDef($container, $con)->addMethodCall('addEventListener', array(array($tag['event']), $lazy ? $id : new Reference($id)));
+                $this->getEventManagerDef($container, $con)->addMethodCall('addEventListener', array(array($tag['event']), $id));
             }
         }
+
+        // replace service container argument of event managers with smaller service locator
+        // so services can even remain private
+        foreach ($listenerClosures as $connection => $closures) {
+            $eventManagerDef = $this->getEventManagerDef($container, $connection);
+            $serviceLocatorId = sprintf($this->managerTemplate, $connection).'.service_locator';
+            $serviceLocatorDef = $container->register($serviceLocatorId, ServiceLocator::class);
```

Comment:
you should use `\Symfony\Component\DependencyInjection\Compiler\ServiceLocatorTagPass::register` instead, which will take care of reusing the same service locator if used multiple times with the same references

Reply:
Thanks was not aware of that :+1: Much simpler.